File: src/euler014.java
/*
    Author - Devesh
    Problem - 013
    Time Complexity - O(n)
*/

import java.util.Scanner;
public class euler014 {
    static int si = 5000005;
    static int ar[] = new int[si];
    static int maxx[] = new int [si];
    public static void main(String[] args) {
        init();
        Scanner inp = new Scanner(System.in);
        int t;
        t = inp.nextInt();
        while(t-- > 0){
            int n;
            n = inp.nextInt();
            System.out.println(maxx[n]);
        }
    }
    static void init() {
        ar[0] = ar[1] = 1;
        for(int i=2; i<si; i++){
            if(ar[i] == 0)
                ar[i] = solve((long)i) - 1;
            // System.out.println(i + " " + ar[i]);
        }
        maxx[1] = 1;
        for(int i=2; i<si; i++){
            if(ar[i] >= ar[maxx[i-1]])
                maxx[i] = i;
            else
                maxx[i] = maxx[i-1];
        }
    }
    static int solve(long n){
        if(n == 1)
            return 1;
        // Return Precomputed Value
        if(n < si) {
            if (ar[(int)n] != 0)
                return ar[(int)n] + 1;
            if(((int)n&1) == 1)
                ar[(int)n] += solve(n*3+1);
            else
                ar[(int)n] += solve((int)n/2);
            return ar[(int)n] + 1;
        }
        else{
            long p = n;
            int j;
            if((p&1) == 1)
                j = solve(n*3+1);
            else
                j = solve(n/2);
            return j + 1;
        }
    }
}

File: C++ Solutions/euler014.cpp
/*
	Author - Devesh
	Problem - 014
	Time Complexity - O(n*depth of recursion)
*/
#include<bits/stdc++.h>
#define ll long
using namespace std;
const int si = 5000005;
ll ar[si];
ll maxx[si];
ll dp(ll n){
    if(n == 1)
        return 1;
    // Return Precomouted Value
    if(n < si){
		// Important -> Would cause segmentation fault since the value of
		// n can go large, so be careful not to query it on array rather
		// put a condition so as to avoid segmentation fault!
        if(ar[n] != 0)
            return ar[n] + 1;
		if(n&1)
			ar[n] += dp(n*3+1);
		else
			ar[n] += dp(n/2);
		return ar[n] + 1;
    }
    else{
        ll p = n, j;
        if(p&1)
            j = dp(n*3+1);
        else
            j = dp(n/2);
        return j + 1;
    }
}
void init(){
    fill(ar, ar+si, 0);
    fill(maxx, maxx+si, 0);
    for(int i=2; i<si; i++){
        if(ar[i] == 0)
            ar[i] = dp(i) - 1;
    //    cout << i << " : " << ar[i] << endl;
    }
    // Store max
    maxx[1] = 2;
    for(int i=2; i<si; i++){
        if(ar[i] >= ar[maxx[i-1]])
            maxx[i] = i;
        else
            maxx[i] = maxx[i-1];
       // cout << ar[i] << " " << ar[maxx[i]] << endl;
    }
}
int main(){
    init();
    int t;
    cin >> t;
    while(t--){
        int n;
        cin >> n;
        cout << maxx[n] <<endl;
    }
}



